### DISCUSSION TITLE
[Access] Paths cannot be queried

### DISCUSSION THREAD

**Discussion by KomlevDm** [payloadcms/payload#2607] [Q&A]
### Link to reproduction

https://github.com/KomlevDm/payload/tree/access-query-issue

### To Reproduce

1. go to `test/_community/collections/Media/index.ts`
2. add query to `MediaCollection.access.read` (`test/_community/collections/Media/prod.ts` and `test/_community/collections/Media/published.ts`)
3. start dev server
4. open `http://localhost:3000/admin/collections/media`
5. page is broken
6. get error `The following paths cannot be queried: _prod, _prod, _status, _status`

[link in test repo](https://github.com/KomlevDm/payload/tree/access-query-issue/test/_community/collections/Media)

### Describe the Bug

Our team has migrated from `1.6.32` to `1.7.4` version. And now we get a global problem (`400 Bad Request`) in admin panel: `The following paths cannot be queried`. We think it's related to `feat: refactors buildQuery to rely on fields instead of mongoose (d187b809d)` in [1.7.0 release](https://github.com/payloadcms/payload/blob/master/CHANGELOG.md#170-2023-04-17). Something very similar was [here](https://github.com/payloadcms/payload/issues/2582), but unfortunately it doesn’t fix current issue. 

<img width="1260" alt="image" src="https://user-images.githubusercontent.com/25402469/235723532-93a3c4ea-803f-4b39-be3f-fa20889156c6.png">

<img width="1827" alt="image" src="https://user-images.githubusercontent.com/25402469/235723575-e0115a9e-fa8e-4276-aa2d-4ef5732a28f5.png">

<img width="632" alt="image" src="https://user-images.githubusercontent.com/25402469/235725647-0d806fa0-e1d7-4771-ab87-fd5cc8a93781.png">

<img width="560" alt="image" src="https://user-images.githubusercontent.com/25402469/235726719-c79af553-53fb-42c3-82eb-897c3a36ae44.png">

<img width="513" alt="image" src="https://user-images.githubusercontent.com/25402469/235728080-5324a822-094e-41d7-9a9d-55dd074d973a.png">

### Payload Version

1.7.4

**Comment by jmikrut** [comment]
Hey @KomlevDm — I can shine some light here for you.

Basically, we have implemented query validation, among other things. Your reproduction repo attempts to query on fields that do not exist in your schema.

I assume you are injecting fields with Mongoose directly or similar, but those fields are not reflected in your Payload collection config(s). 

To fix this, you need to make sure that you have fields that reflect anything that you query on. If you don't want them to show up in the admin UI, but you want them to be available to query on, you can set these fields to `admin.disabled` to hide them entirely from admin UI, but then your queries will once again work.

Regarding the `_status` field, same thing there - your Media collection does not have `versions.drafts: true`, so you have no `_status` field, and thus can't query on that field.

I will convert this to a discussion because everything is functioning as-intended, but I am happy to continue helping!

**Reply by KomlevDm** [reply]
Yes, but we just want to check documents for the presence of the marker field (pre-prod). Therefore, we ask if the document has a marker field. We want to add this marker field only when creating content by certain roles, that is, this marker will not always exist in the document.

Therefore, we do not want to hide it in UI, but we want it to be somewhere, and somewhere not.

```ts
  or: [
    {
      _preProd: {
        exists: false,
      },
    },

    {
      '_preProd.isDeployed': {
        equals: true,
      },
    },
  ],
```

**Reply by KomlevDm** [reply]
We implemented the addition of this field through the plugin and hooks, and apparently the payload does not know about it, as I understood from the answer above.

<img width="685" alt="image" src="https://user-images.githubusercontent.com/25402469/235732766-7d1eca20-e882-4717-913a-c70cbfdcacfc.png">


<img width="450" alt="image" src="https://user-images.githubusercontent.com/25402469/235732640-b1f83eeb-65f1-4f5f-ae2b-25d8c31ebb19.png">

**Reply by jmikrut** [reply]
Any fields you add with a plugin are completely respected with our query validation. But the problem is, is that field is _**hidden**_. Change it to `disabled` instead of `hidden` and you will then be able to query on it!

The reason why we are now restricting hidden fields from being queried on is because that could actually pose as a security vulnerability. More info here:

https://github.com/payloadcms/payload/security/advisories/GHSA-35jj-vqcf-f2jf

**Reply by DanRibbens** [reply]
Hold up, `admin.hidden` is not the same as `hidden` on the field. 

@jmikrut the code example is using `admin.hidden` and that shouldn't make any difference for querying.

**Reply by KomlevDm** [reply]
Thank you for explaining the reason for creating a validation - here I agree and will look towards `disabled`.

But about question on the `_status` for work with drafts.
That is, now there is no way to check any document when requesting for the presence of a field (_status) if the collection does not have versioning. It's just that our media collection does not have versioning and there are some other collections that do not need buttons (Save Draft and Publish changes) that come with versioning. Can we somehow check all this at one point in the plugin (as in version 1.6.32 before validation)?

**Reply by KomlevDm** [reply]
Guys, I re-read everything you wrote and what I could find in the Issues, PRs etc ([1](https://github.com/payloadcms/payload/issues/2592), [2](https://github.com/payloadcms/payload/issues/2582), [3](https://github.com/payloadcms/payload/pull/2599), [4](https://github.com/payloadcms/payload/commit/d187b809d7b0b37465b15d0b4b17702be521f630)). Here is the result:

1. Use `hidden` property for group field (based on the docs and [PR](https://github.com/payloadcms/payload/pull/2599), that _will still be saved to the database, but will not appear in any API or the Admin panel_)
   <img width="471" alt="image" src="https://user-images.githubusercontent.com/25402469/235778194-76b64a3e-1c6f-4250-8641-9711667b84a1.png">
   <img width="1253" alt="image" src="https://user-images.githubusercontent.com/25402469/235779892-38ea95b5-7686-4e35-b6bd-cee8cfe460e4.png">

2. Explicitly add the configuration of this group field to all collections in the plugin (due to the appearance of validation from version `1.7.0`)
3. I have not yet been able to solve the issue with the `_status` field, I continue to believe that the current implementation through versioning is not exactly elegant. For example, media or other collections that do not need versioning in the app should simply be able to allow access based on the absence of `_status` field. As I see that in order for the old logic (which, by the way, was in the docs) to work, the `_status` field will also need to be explicitly set for all collection configurations. And it would be great to be able to do it easier, because the payload can work with versioning out of box and it seems redundant to add a separate field (e.g [isPublic](https://payloadcms.com/docs/access-control/collections#read), etc.).
   <img width="687" alt="image" src="https://user-images.githubusercontent.com/25402469/235795958-3dcfabfd-0b46-4790-aaec-c54c0661190c.png">

`But` I still have to add `afterRead` hook to remove the field object completely, because it still leaks, even though `hidden`:
<img width="414" alt="image" src="https://user-images.githubusercontent.com/25402469/235782846-4a05fa1a-b279-4661-b59a-452888a1fb39.png">

`The same problem` now appears in the case of a document that did not need a field before, but now it has it in the configuration and it gets into the db as a vestige of an empty object:
<img width="331" alt="image" src="https://user-images.githubusercontent.com/25402469/235785984-70545b11-06d7-45af-a5b7-33e5bac8471a.png">

`But` for an access query, this is solved strangely, but probably logically:
```ts
    {
      _preProd: {
        equals: {},
      },
    }
```

**Comment by thgh** [comment]
I had the same issue, it turned out to be related to a relationship field that had a read access function that returned an invalid query.

The error details are quite limited though.

**Reply by ccssmnn** [reply]
Thank you @thgh ! Your hint to the access function pointed me to the right direction.

I had the same issue. In my case, the access function returned a filter that included a condition on the `user` field. But the collection did not have a user field.

```
[17:09:46] ERROR (payload): QueryError: The following path cannot be queried: user
    at validateQueryPaths (/myproject/node_modules/payload/src/database/queryValidation/validateQueryPaths.ts:91:13)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async find (/myproject/node_modules/payload/src/collections/operations/find.ts:138:7)
    at async getEntityDoc (/myproject/node_modules/payload/src/utilities/getEntityPolicies.ts:58:32)
    at async createAccessPromise (/myproject/node_modules/payload/src/utilities/getEntityPolicies.ts:119:40)
    at async /myproject/node_modules/payload/src/utilities/getEntityPolicies.ts:245:5
    at async /myproject/node_modules/payload/src/utilities/getEntityPolicies.ts:228:5
    at async /myproject/node_modules/payload/src/utilities/getEntityPolicies.ts:228:5
    at async getEntityPolicies (/myproject/node_modules/payload/src/utilities/getEntityPolicies.ts:227:3)
    at async docAccess (/myproject/node_modules/payload/src/collections/operations/docAccess.ts:43:20)
```
